### PR TITLE
Host discovery

### PR DESCRIPTION
**Pull request by ghazanfarttu**
This is initial update of "Host" discovery module which intends to discover host states locally (using in-band mechanism). This revision implements on-host discovery of CPU thermal state.

### CODE REVIEW

File: modules/hostdiscovery/hostdiscovery.go [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+	tempVal := hostDisc.ReadCPUTemp()
+	cpuTempObj.CPUTemp = tempVal
+
+	return cpuTempObj
+}
+
+// ReadCPUTemp function reads the CPU thermal sensor
+func (hostDisc *HostDisc) ReadCPUTemp() int {
+
+	tempSensorPath := "/sys/devices/virtual/thermal/thermal_zone0/temp"
```
Comment (ghazanfarttu) [inline modules/hostdiscovery/hostdiscovery.go]: make it configurable

File: modules/hostdiscovery/hostdiscovery.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+		return 0
+	}
+	cpuTempInt, err := strconv.Atoi(strings.TrimSuffix(string(cpuTemp), "\n"))
+
+	if err != nil {
+		hostDisc.api.Logf(lib.LLERROR, "String to Int conversion failed: %v", err)
+		return 0
+	}
+
+	return cpuTempInt
```
Comment (ghazanfarttu) [inline modules/hostdiscovery/hostdiscovery.go]: send if value changes

Comment (ghazanfarttu) [reply modules/hostdiscovery/hostdiscovery.go]: I reflected this change before generating temperature discovery event.

File: modules/hostdiscovery/hostdiscovery.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+	return ip
+}
+
+// Discovers state of the CPU based on CPU temperature thresholds
+func lambdaStateDiscovery(v CPUTempObj) (string, int) {
+	cpuTemp := v.CPUTemp
+	cpuTempState := thpb.HostThermal_CPU_TEMP_NONE
+
+	if cpuTemp <= 3000 || cpuTemp >= 70000 {
```
Comment (ghazanfarttu) [inline modules/hostdiscovery/hostdiscovery.go]: Make normal, high and critical thresholds configurable.

Comment (ghazanfarttu) [reply modules/hostdiscovery/hostdiscovery.go]: resolved by adding these fields into protobuff

File: extensions/HostThermal/HostThermal.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+import (
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/hpc/kraken/core"
+	pb "github.com/hpc/kraken/extensions/HostThermal/proto"
+	"github.com/hpc/kraken/lib"
+)
+
+//go:generate protoc -I ../../core/proto/include -I proto --go_out=plugins=grpc:proto proto/Thermal.proto
```
Comment (kpelzel) [inline extensions/HostThermal/HostThermal.go]: Correct this comment

File: extensions/HostThermal/proto/HostThermal.proto [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ syntax = "proto3";
+ package proto;
+
+message HostThermal {
+     enum hostThermalState {
+        CPU_TEMP_NONE = 0;
+        CPU_TEMP_NORMAL = 1;
+        CPU_TEMP_HIGH = 2;
+        CPU_TEMP_CRITICAL = 3;
+        //TEMP_WAIT = 1; // Thermal request sent to node
```
Comment (kpelzel) [inline extensions/HostThermal/proto/HostThermal.proto]: Probably should remove this comment if it's not being used.

File: modules/hostthermaldiscovery/hostthermaldiscovery.go [hunk 6]
```diff
@@ -0,0 +1,273 @@
+/* HostThermalDiscovery.go: This module performs CPU Thermal discovery of HPC nodes through in-band (through OS) mechanism.
+ *
+ * Author: Ghazanfar Ali, ghazanfar.ali@ttu.edu; Kevin Pelzel <kevinpelzel22@gmail.com>; J. Lowell Wofford <lowell@lanl.gov>
+ *
+ * This software is open source software available under the BSD-3 license.
+ * Copyright (c) 2018, Triad National Security, LLC
+ * See LICENSE file for details.
+ */
+
+//go:generate protoc -I ../../core/proto/include -I proto --go_out=plugins=grpc:proto proto/pxe.proto
```
Comment (kpelzel) [inline modules/hostthermaldiscovery/hostthermaldiscovery.go]: Correct this comment

File: modules/hostthermaldiscovery/proto/hostthermaldiscovery.proto [hunk 7]
```diff
@@ -0,0 +1,29 @@
+/* hostdiscovery.proto: describes the hostdiscovery object
```
Comment (kpelzel) [inline modules/hostthermaldiscovery/proto/hostthermaldiscovery.proto]: Correct this comment

**Pull request merged**